File: 1001-2000/1004. Max Consecutive Ones III.java
// Medium
// Sliding Window
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/max-consecutive-ones-iii/

class Solution {
    public int longestOnes(int[] nums, int k) {
        int left = 0, maxLen = 0, zero = 0;
        for (int right = 0; right < nums.length; right++) {
            if (nums[right] == 0) {
                zero++;
            }
            while (zero > k) {
                if (nums[left] == 0) {
                    zero--;
                }
                left++;
            }
            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }
}

File: 1001-2000/1004. Max Consecutive Ones III.py
# Time:O(n), Space:O(1)

from typing import List

class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        n = len(nums)
        zeros = 0
        max_len = 0
        left = 0

        for right in range(n):
            if nums[right] == 0:
                zeros += 1
            
            while zeros > k:
                if nums[left] == 0:
                    zeros -= 1
                left += 1
            
            max_len = max(max_len, right - left + 1)
        
        return max_len
